numpy/core/src/multiarray/descriptor.c

```diff
@@ -262,7 +262,24 @@ _convert_from_tuple(PyObject *obj, int align)
     /*
      * We get here if _try_convert_from_inherit_tuple failed without crashing
      */
-    if (PyDataType_ISUNSIZED(type)) {
+    if (NPY_DT_is_user_defined(type)) {
+        /* interpret next item as a typesize */
+        PyObject* itemsize = PyTuple_GET_ITEM(obj, 1);
+
+        PyArray_DTypeMeta* DType = NPY_DTYPE(type);
+
+        PyObject *res = PyObject_CallFunctionObjArgs(
+            (PyObject *)DType, itemsize, NULL);
+
+        Py_DECREF(type);
+
+        if (res == NULL) {
+            return NULL;
+        }
+
+        return (PyArray_Descr *)res;
+    }
+    else if (PyDataType_ISUNSIZED(type)) {
```

Comment:
In principle we could now move all of this into Python, but I am happy to skip it here.  (To simplify the C-code to always get a dtype instance.)